### PR TITLE
[Serializer] Add a ScalarDenormalizer

### LINKED ISSUE #38983: [Serializer] Impossible to use UnwrappingDenormalizer with scalar denormalization
**Symfony version(s) affected**: 5.1.8

**Description**
It's impossible to denormalize into a scalar value when using the unwrap path feature, because the scalar value check is done before applying the UnwrappingDenormalizer.

**How to reproduce**
```php
use Symfony\Component\Serializer\Normalizer\UnwrappingDenormalizer;

$result = $serializer->deserialize(
    '{"baz": "test"}',
    'string',
    'json',
    [UnwrappingDenormalizer::UNWRAP_PATH => '[baz]']
);

// expected result
// $result === "test"

// actual result
// Symfony\Component\Serializer\Exception\NotNormalizableValueException: Data expected to be of type "string" ("array" given).
```

**Possible fix**
Introduce a ScalarDenormalizer instead of checking the target type inside the Serializer class.

### PR DESCRIPTION
**Pull request by camilledejoye**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | no
| License       | MIT
| Doc PR        | TBD

The goal is to use a denormalizer to handle scalar values instead of doing it directly in the `Serializer` class.
This allows to:
* Execute other normalizers first (based on their priority), e.g. `UnwrappingDenormalizer`
* Add more logic without overwhelming the `Serializer` code, e.g. specific behavior based on the format

Todos:
- [ ] Doc PR ? Not sure where though, the doc mainly mentions normalizers doing both operations
- [ ] New PR to register `ScalarDenormalizer` as a service in `FrameworkBundle`
- [ ] New PR to remove duplication with `AbstractObjectDenormalizer` for more clarity

### CODE REVIEW

File: src/Symfony/Component/Serializer/Serializer.php [hunk 1]
```diff
@@ -189,14 +182,6 @@ public function normalize($data, string $format = null, array $context = [])
...
     public function denormalize($data, string $type, string $format = null, array $context = [])
     {
-        if (isset(self::SCALAR_TYPES[$type])) {
-            if (!('is_'.$type)($data)) {
-                throw new NotNormalizableValueException(sprintf('Data expected to be of type "%s" ("%s" given).', $type, get_debug_type($data)));
-            }
-
-            return $data;
-        }
-
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Serializer.php]: This is a breaking change because code that relied on that logic now has to configure an additional normalizer.

File: src/Symfony/Component/Serializer/Serializer.php [hunk 2]
```diff
@@ -179,7 +172,7 @@ public function normalize($data, string $format = null, array $context = [])
             throw new NotNormalizableValueException(sprintf('Could not normalize object of type "%s", no supporting normalizer found.', get_debug_type($data)));
         }
 
-        throw new NotNormalizableValueException('An unexpected value could not be normalized: '.(!\is_resource($data) ? var_export($data, true) : sprintf('%s resource', get_resource_type($data))));
+        throw new NotNormalizableValueException('An unexpected value could not be normalized: '.(!\is_resource($data) ? var_export($data, true) : sprintf('"%s" resource', get_resource_type($data))));
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Serializer.php]: This is an unrelated change and should be reverted.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php [hunk 3]
```diff
@@ -558,47 +558,51 @@ public function testNormalizeScalarArray()
...
         $this->assertSame('  spaces  ', $serializer->deserialize('"  spaces  "', 'string', 'json'));
         $this->assertSame('@Ca$e%', $serializer->deserialize('"@Ca$e%"', 'string', 'json'));
     }
 
     public function testDeserializeLegacyScalarType()
     {
-        $this->expectException(LogicException::class);
-        $serializer = new Serializer([], ['json' => new JsonEncoder()]);
+        $this->expectException(NotNormalizableValueException::class);
+        $serializer = new Serializer([new ScalarDenormalizer()], ['json' => new JsonEncoder()]);
```
Comment (julienfalque) [inline src/Symfony/Component/Serializer/Tests/SerializerTest.php]: These tests demonstrate that now you need to pass a `ScalarDenormalizer` instance to the `Serializer` to be able to deserialize scalar values. Shouldn't this be considered a BC break and require a compatibility layer?

**Review by julienfalque — COMMENTED** [review]
:+1:

File: src/Symfony/Component/Serializer/Serializer.php (hunk 2 above)
Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Serializer.php]: This was required by fabbot.it

File: src/Symfony/Component/Serializer/Serializer.php (hunk 1 above)
Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Serializer.php]: You're right, when I started working on this I thought this change was a lot more recent and I forgot to consider the impact.
I'll try to find a way to revert that so that the PR won't break anything but maybe we should deprecate this use case ?

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php (hunk 3 above)
Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Tests/SerializerTest.php]: Indeed, I will resolve this conversation since it a duplicate of https://github.com/symfony/symfony/pull/39140#discussion_r528387266

File: src/Symfony/Component/Serializer/Serializer.php (hunk 1 above)
Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Serializer.php]: I put the code back and added a condition to only use it the the `ScalarDenormalizer` is not registered.
I also added a deprecation and tried to add the necessary information in the different files as mentioned in the conventions.

File: src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\PropertyInfo\Type;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Exception\NotNormalizableValueException;
+
+final class ScalarDenormalizer implements DenormalizerInterface, CacheableSupportsMethodInterface
```
Comment (julienfalque) [inline src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: Maybe register the new denormalizer in `src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.php`?

File: src/Symfony/Component/Serializer/Serializer.php [hunk 5]
```diff
@@ -189,7 +184,9 @@ public function normalize($data, string $format = null, array $context = [])
      */
     public function denormalize($data, string $type, string $format = null, array $context = [])
     {
-        if (isset(self::SCALAR_TYPES[$type])) {
+        if (isset(ScalarDenormalizer::SCALAR_TYPES[$type]) && !$this->hasScalarDenormalizer()) {
+            trigger_deprecation('symfony/serializer', '5.2', 'Denormalizing scalar values without registering the "%s" is deprecated.', ScalarDenormalizer::class);
```
Comment (julienfalque) [inline src/Symfony/Component/Serializer/Serializer.php]: SemVer forbids deprecating things in patch releases so this PR should be merged into branch `5.2` if it's still ok, `5.x` otherwise.

Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Serializer.php]: Yes, not even sure the implementation will be accepted.
That's why I tagged the deprecation on `5.2`, the milestone will be change later if this goes any further.

File: src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php (hunk 4 above)
Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: I reverted it, it's a good point but I think it should be done in a separate PR.
The test are failing otherwise since the `ScalarDenormalizer` does not exist in the version of the code retrieved by composer.

File: src/Symfony/Component/Serializer/Serializer.php [hunk 6]
```diff
@@ -299,6 +296,21 @@ private function getDenormalizer($data, string $class, ?string $format, array $c
...
     }
 
+    private function hasScalarDenormalizer(): bool
+    {
+        if (null !== $this->hasScalarDenormalizer) {
+            return $this->hasScalarDenormalizer;
+        }
+
+        foreach ($this->normalizers as $normalizer) {
+            if ($normalizer instanceof ScalarDenormalizer) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Serializer.php]: What about using `supportsNormalization` instead?

This may fail if someone decorated the normalizer (e.g for tracking purposes).

Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Serializer.php]: Indeed, checking the type was a bad idea just to avoid some possibly costly computation in a custom normalizer but it does more wrong than good.
I think I will go back to my first idea and use the `Seriliazer::getDenormalizer()` method instead since it's already cached.

**Comment by julienfalque** [comment]
There seems to be similar logic in `AbstractObjectNormalizer::validateAndDenormalize()`, maybe we can rely on this new denormalizer instead?

**Review by dunglas — COMMENTED** [review]
LGTM, but I agree with @julienfalque, maybe could we get rid of the duplicated logic.

**Comment by camilledejoye** [comment]
> There seems to be similar logic in `AbstractObjectNormalizer::validateAndDenormalize()`, maybe we can rely on this new denormalizer instead?

I agree too but I was thinking on doing it in a separate PR.
This normalizer never used the serializer to normalize scalar values, if I do it right now we will have another BC break because if the `ScalarDenormlizer` is not register then deserializing `'3'` to a float won't work anymore.
The only way to do it is to move the currently duplicated logic from the `AbstractObjectNormalizer` into the deprecated handling of scalar values in the `Serializer`.
So we would still have a duplication, but at least it would be removed in `6.0` when deleted the deprecated code.

The main reason I wanted to do it in another ticket is because this normalizer only call `Serializer::denormalize()` for properties that are an object or an array of objects.
This introduce a bug when an object has property which is an array of integers for instance, the only check done is that the data is an array but it could still contains strings and the denormalizer would not complains, which would put the target in an invalid state.
So I think it would be better to have another bugfix that will make `AbstractObjectNormalizer::validateAndDenormalize()` use the serializer to denormalize everything and not only inner objects, which will fix the issue mentioned above.

Well, now that I'm looking at it again I think the logic applied when denormalizing from XML or CSV should also be added into the `ScalarDenormalizer` ...

**Comment by julienfalque** [comment]
Refactoring `AbstractObjectNormalizer` in another iteration makes sense to me.

File: src/Symfony/Component/Serializer/Tests/SerializerTest.php [hunk 7]
```diff
@@ -606,13 +614,127 @@ public function testDeserializeScalarArray()
...
+        $this->assertSame(3.0, $serializer->deserialize('3', 'float', 'json')); // '3' === json_encode(3.0)
+        $this->assertSame('  spaces  ', $serializer->deserialize('"  spaces  "', 'string', 'json'));
+        $this->assertSame('@Ca$e%', $serializer->deserialize('"@Ca$e%"', 'string', 'json'));
+    }
+
+    public function testDeserializeScalarFromXml()
+    {
+        $serializer = new Serializer([new ScalarDenormalizer()], ['xml' => new XmlEncoder()]);
+
+        $deserialize = fn ($value, string $type) => $serializer->deserialize("<?xml version='1.0'?>\n<tag>$value</tag>", $type, 'xml');
```
Comment (andrew-demb) [inline src/Symfony/Component/Serializer/Tests/SerializerTest.php]: Symfony 5.2 supports PHP ^7.2, where short arrow functions are not available.

File: src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+use Symfony\Component\Serializer\Encoder\CsvEncoder;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Encoder\XmlEncoder;
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Exception\NotNormalizableValueException;
+
+final class ScalarDenormalizer implements DenormalizerInterface, CacheableSupportsMethodInterface
+{
+    public const SCALAR_TYPES = [
+        Type::BUILTIN_TYPE_INT => true,
```
Comment (andrew-demb) [inline src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: Usage constants from `symfony/property-info` will require to be installed component always.
`symfony/serializer` uses property info as an optional dependency for now, isn't it?

Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: You're right, which is strange since both `symfony/property-access` and `symfony/property-info` are used in other places like `Serializer` and `AbstractObjectNormalizer` classes...

Comment (stof) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: the property-info component is optional in AbstractObjectNormalizer

Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: I think I don't understand what you mean by optional then:
https://github.com/symfony/symfony/blob/49eafee03323253328d76ec9cb4c515f94d2f202/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php#L17

Does not that mean that without the `PropertyInfo` component the class will not be found ?

Comment (stof) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: a use statement does not require the class to exist. Resolving a use statement happens purely at compile-time, without loading the class.
And in AbstractObjectNormalizer, usages of `Type` constants happen only in code paths where the type extractor is not null (at which point we can be sure that the component is installed)

Comment (camilledejoye) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: Good to know, so even the type hint for type extractor does not trigger the check for the class is we pass null I guess.
Thanks for the explanation.

Comment (stof) [reply src/Symfony/Component/Serializer/Normalizer/ScalarDenormalizer.php]: yeah, an `instanceof` check (and typehints are applying such checks) does not trigger any autoloading in PHP. They check the passed object.

**Comment by derrabus** [comment]
We cannot do a deprecation in a bugfix release. If this PR indeed fixes a bug, we would need to split it:
* A bugfix for the 5.1 branch.
* A feature PR including the deprecation for the 5.x branch, targeting the 5.3 release.

**Comment by camilledejoye** [comment]
I understand and after fixing the different returns this PR is more of an improvement than a bugfix.

It might be possible by checking if a denormalizer is available before trying to denormalize the scalar value, as done in this implementation.

How should I proceed:
* Keeping this PR for the improvement so that we kept the history and creating a new one for the bugfix (needs to update labels/milestone) ?
* Duplicate this branch to create another PR for the feature and forced push the bugfix only to this PR ?
* Close this PR and create two new ones, mentioning this one into the improvement for reference ?

**Comment by derrabus** [comment]
> Keeping this PR for the improvement so that we kept the history and creating a new one for the bugfix

I think, that's the best way.

**Comment by soyuka** [comment]
Also if we fix the UnwrappingDenormalizer I think that its priority needs to be reviewed, it's the only one in symfony to be `> 0`. The highest is `-890` right now, this one could be at `-700`, IMO it'd be more consistent then using `1000`. I'm not sure how these priorities are decided though.  

```
Symfony Container Services Tagged with "serializer.normalizer" Tag
==================================================================

 ----------------------------------------------------------- ---------- -----------------------------------------------------------------------------------------------------
  Service ID                                                  priority   Class name
 ----------------------------------------------------------- ---------- -----------------------------------------------------------------------------------------------------
  serializer.denormalizer.unwrapping                          1000       Symfony\Component\Serializer\Normalizer\UnwrappingDenormalizer
  serializer.normalizer.object                                -1000      Symfony\Component\Serializer\Normalizer\ObjectNormalizer
  serializer.normalizer.datetimezone                          -915       Symfony\Component\Serializer\Normalizer\DateTimeZoneNormalizer
  serializer.normalizer.dateinterval                          -915       Symfony\Component\Serializer\Normalizer\DateIntervalNormalizer
  serializer.normalizer.data_uri                              -920       Symfony\Component\Serializ er\Normalizer\DataUriNormalizer
  serializer.normalizer.datetime                              -910       Symfony\Component\Serializer\Normalizer\DateTimeNormalizer
  serializer.normalizer.json_serializable                     -900       Symfony\Component\Serializer\Normalizer\JsonSerializableNormalizer
  serializer.normalizer.problem                               -890       Symfony\Component\Serializer\Normalizer\ProblemNormalizer
  serializer.denormalizer.array                               -990       Symfony\Component\Serializer\Normalizer\ArrayDenormalizer
  serializer.normalizer.constraint_violation_list             -915       Symfony\Component\Serializer\Normalizer\ConstraintViolationListNormalizer
 ----------------------------------------------------------- ---------- -----------------------------------------------------------------------------------------------------
```

**Comment by camilledejoye** [comment]
My guess is that the priority is so high because this denormalizer does not rely on the type and must always be called first.
I didn't check but assuming the default priority for any custom denormalizer is `0` then the `UnwrappingDenormlizer` must indeed have a priority `> 0`.

The bug that was reported occurred because scalar values are handled directly in the serializer and therefore it has the priority over the `UnwrappingDenormalizer`.

**Comment by camilledejoye** [comment]
Little update, while trying to remove the duplication between the `ScalarDenormalizer` and the `AbstractObjectNormalizer` I realized there is a few small bugs here and there that makes it difficult so I was wondering if it might not be beneficial to solve them first and delay this feature until then.

First, the scalar denormalization is done directly in the `Serializer` and use a relatively naive approach: it does not handle the specific cases dealt with in the `AbstractObjectNormalizer`.
For example, deserializing `'3'` into a `float` type will fail but when decoding from JSON it should be accepted because `'3' === json_encode(3.0)` (there is other cases for XML and CSV).

Second, the `AbstractObjectNormalizer` only delegate the denormalization process when dealing with properties which types are `object` or `object[]`. For all other cases (builtin types) it uses the `is_<type>()` PHP function to validate it and if the data is valid returns it.
Which means it handles scalars and arrays internally, it's not a problem for scalar denormalization since it does a better job than the `Serializer` but it does not validate arrays properly.
If a property is of type `int[]` and the data to the denormalize is `['string']` the `AbstractObjectNormalizer` will consider it valid (it's an array) and will try to set the value onto the object.

To summarize there is two issues:
* The scalar denormalization in the `Serializer` is not smart enough
* The `AbstractObjectNormalizer` does not properly validate arrays

What we could do is:
1. A PR to fix the scalar denormalization bugs: duplicate the logic from the `AbstractObjectNormalizer` into the `Serializer` (target 5.1)
2. A PR to make the `AbstractObjectNormalizer` delegate all denormalization to the serializer (target 5.1 if consider a hotfix but do not deprecate using it without the delegation, 5.x if consider a feature and then add deprecation)
3. Update this PR to introduce the `ScalarDenormalizer` and deprecate the scalar denormalization within the `Serializer` (also deprecate using the fallback in the `AbstractObjectNormalizer` if step `2.` was done as a bugfix)

But the step `2.` would trigger two BC breaks:
* `AbstractObjectNormalizer` will not be usable without setting the `serializer` property, which is only required during denormalization to handle objects currently
* Denormalizing arrays without registering the `ArrayDenormalizer` won't work anymore

To void this we need to find a way to fallback to the current behavior if we can't denormalize because:
* The `serializer` property is not an instance of `DenormalizerInterface`
* There was nos denormalizer available: `ArrayDenormalizer` not register or for the builtin types `iterable`, `resource` and `callable`.

It should be doable by adding an new exception like `NoDenormalizerFoundForValueException` extending `NotNormalizableValueException`.
The problem with this solution is that we ends up duplicating the code responsible to handle scalar values 3 times.

I would like to validate how you would suggest to process before starting anything on my side.
Should I create a dedicated issue to talk about this ? If so what kind of issue should it be ?

**Comment by soyuka** [comment]
Idk if this is relevant but you should also consider https://github.com/symfony/symfony/issues/30818 and the work around it.
